Gradata/tests/test_cloud_materializer.py

```
@@ -0,0 +1,404 @@
+            result = materialize(sorted(shuffled, key=lambda e: e["ts"])).rules
+            assert result == baseline
+
+    def test_convergence_with_injected_conflicts(self) -> None:
+        """Even when some keys enter Tier 2 hold, replay order is irrelevant."""
+        rng = random.Random(4242)
+        pool = self._build_pool(rng, n_keys=4, per_key=3)
+        # Inject a large-drift event per key → forces a conflict hold.
+        for k in range(4):
+            pool.append(
+                _evt(
+                    ts=f"2026-04-20T00:30:00Z",
```

Ruff (0.15.10)

[error] f-string without any placeholders

Remove extraneous `f` prefix

(F541)